WINC-1007: Windows CSI Proxy enhancement proposal (#1372)
openshift/enhancements: enhancements/windows-containers/csi-proxy.md

---
title: csi-proxy
authors:
  - "@sebsoto"

> reviewers:

Reviewer: There's some error with this line in the doc preview: https://github.com/sebsoto/enhancements/blob/csiProxy/enhancements/windows-containers/csi-proxy.md

Author: Thats odd i'm not sure why thats happening

Author: Fixed

storage drivers, [which are in the process of being deprecated and removed.](https://kubernetes.io/blog/2022/09/26/storage-in-tree-to-csi-migration-status-update-1.25/)

This will be done by running [CSI Proxy](https://github.com/kubernetes-csi/csi-proxy) as a Windows service on each
Windows node. csi-proxy is a pre-requisite for CSI node driver pods to interact with the underlying Windows storage
APIs. With csi-proxy running on a cluster's Windows nodes, cluster administrators are free to deploy the CSI storage
drivers of their choice as a Windows DaemonSet, enabling Windows pods on each node to have access to storage through
PersistentVolumeClaims.

## Motivation

Windows nodes on OpenShift have had persistent storage available through the use of the in-tree storage drivers.
Persistent storage is a necessary feature for production workloads on Kubernetes, and it is important that this
functionality continue to work, even with the in-tree drivers being removed.

### User Stories

* As a developer using OpenShift, I want the Windows workloads I create to be able to share storage between workloads,
  and have that storage persist when pods are removed and re-created.

> ### Goals

Reviewer 2: It should be explicitly called out somewhere if using the volume by Linux and Windows Pods is a requirement or not. It might be tricky due to filesystem support, not sure if / how we support NTFS on Linux or ext4/xfs on Windows.

Reviewer 3: @jsafrane this has been called out in non-goals now. Does that suffice?

Reviewer 2: Yes

* Windows nodes configured by WMCO fulfill all the pre-requisites to run CSI node drivers.

### Non-Goals

* The Windows CSI storage drivers will not be distributed as part of OpenShift.
  * At this time, we are not able to build or ship Windows images due to support restrictions.

## Proposal

The [windows-machine-config-operator](https://github.com/openshift/windows-machine-config-operator/)(WMCO),
will ensure that csi-proxy is running on the Windows nodes it configures. This will be done through the normal
workflow for adding a service to the expected service list.

The csi-proxy binary will be baked into the WMCO image. WMCO will copy the binary to each node at configuration
time. A new item will be added to the windows-services ConfigMap, instructing the windows-instance-config-daemon
to ensure that csi-proxy is always running as a Windows service on each node.

With the nodes in this state, users are free to follow upstream documentation to deploy the CSI node drivers
of their choice on the Windows nodes in the cluster.

No support from Red Hat will be given for issues occurring with the upstream drivers, as they are not being distributed

> as part of OpenShift.

Reviewer: Will we eventually support these drivers?

Author: Thats unclear for now, I'd prefer not to speculate about it in the enhancement.

Reviewer 4: We are still going to ship and support drivers that are part of OCP installs right? Such as - EBS, vSphere etc?

Reviewer 4: Okay, so reading non-goals it appears that we are not even going to support drivers which are part of standard OCP install on Linux for example? What happens on migration then? Am I reading this right?

Author: Thanks for the comments @gnufied 
I've updated this doc to clarify that we will still support issues with the CSI driver controller shipped by openshift.

### Workflow Description

1. The cluster administrator will identify what CSI storage drivers the cluster requires for its workloads.

> 2. The cluster admin can follow upstream documentation to deploy the CSI node drivers on its Windows nodes.

Reviewer 3: I would say something like `Based on the requirements...` to connect the two points.

### API Extensions

N/A

### Risks and Mitigations

If there are any security issues present in the CSI driver images deployed by the user, they will not be fixed

> by upgrading OCP. It is up to the user to stay informed and update their images as necessary.

Reviewer 3: - Isn't there a risk of an upgrade potentially breaking a driver?
- What about a mismatch in versions between Linux and Windows? You should link to the version skew section here.

### Drawbacks

> The main drawback to this approach is it puts a lot of work on the cluster administrator. As they will be required

Reviewer 3: ```suggestion
The main drawback to this approach is it puts the onus on the cluster administrator
```

to deploy and maintain the Windows CSI DaemonSet themselves, and will not be given support for any issues that may
come up, which do not specifically deal with the csi-proxy Windows service. This may cause an increased load
on both support and developers, as support issues coming in will need to be properly filtered out based on the root
cause of the problem.

## Design Details

### Test Plan

WMCO already has an extensive end to end test suite. To facilitate testing the additions described in this enhancement,

> the e2e tests will be augmented to deploy a CSI node driver DaemonSet across Windows nodes, just as a user would. The

Reviewer 5: Which platform(s) and driver(s) will we test?

e2e tests would then validate that a Windows pod is able to mount a volume backed by the driver.

### Graduation Criteria

#### Dev Preview -> Tech Preview

N/A

#### Tech Preview -> GA

This feature is targeted for WMCO 9.0.0 in OpenShift 4.14.

#### Removing a deprecated feature

No action required.

### Upgrade / Downgrade Strategy

[When upgrading cluster versions, in-tree driver Volumes will automatically be updated to use the appropriate CSI node
drivers by the cluster](https://docs.openshift.com/container-platform/4.12/storage/container_storage_interface/persistent-storage-csi-migration.html).

If users are making use of the Linux CSI node drivers on the cluster, and not deploying all parts of storage

> themselves, some extra care must be had when upgrading a cluster.

Reviewer 3: What is this extra care?

Author: The extra care is what is written in the rest of this section

Reviewer 3: Then please add `as described below`.

Before upgrading, users should check the OCP release notes to determine how the Linux CSI drivers have changed, and
if they need to use a new Windows CSI node driver image. If so, the Windows DaemonSet should be updated once the
cluster upgrade is complete.

WMCO does not support downgrades.

### Version Skew Strategy

If the CSI driver pods included in OCP are used as the driver controller and Linux node drivers, while the Windows

> node drivers use the upstream images, there is going to be a mismatch. This should be mitigated by the cluster

Reviewer 3: You are guaranteeing a mismatch. Is that always the case? You could potentially end up using the same version.

administrator ensuring that the upstream images they use are as close to a version match as possible with the OCP
distributed drivers.

To eliminate this issue, the cluster administrator can choose to use an OCP cluster with no CSI drivers installed.
They would then install both the Windows and Linux CSI pods and required resources themselves, using upstream
documentation. This is only an option if they only plan on using storage on Windows, as doing this will prevent the
user from getting support from Red Hat for any storage issues they may have, both Linux and Windows.

> 

Reviewer 3: You should also mention using a separate cluster for Windows container workloads.

themselves, users will have to ensure the following:
Before upgrading, users should check the OCP release notes to determine how the Linux CSI drivers have changed, and
if they need to use a new Windows CSI node driver image. If so, the Windows DaemonSet should be updated once the
cluster upgrade is complete. More on this can be found in the [version skew](#version-skew-strategy) section.

WMCO does not support downgrades.

### Version Skew Strategy

If the CSI driver pods included in OCP are used as the driver controller and Linux node drivers, while the Windows
node drivers use the upstream images, there may be a mismatch. This should be mitigated by the cluster administrator
ensuring that the upstream images they use are as close to a version match as possible with the OCP distributed
drivers.

To eliminate this issue, the cluster administrator can choose to do the following:
1) When only Windows workloads need access to storage, an OCP cluster with no CSI drivers installed can be
   used. The cluster admin would then deploy both the Windows and Linux CSI drivers and required resources,

>    by following upstream documentation.

Reviewer 3: Why would the customer need Linux CSI drivers if they only need Windows storage?

Windows node. csi-proxy is a pre-requisite for CSI node driver pods to interact with the underlying Windows storage
APIs. With csi-proxy running on a cluster's Windows nodes, cluster administrators are free to deploy the CSI storage
drivers of their choice as a Windows DaemonSet, enabling Windows pods on each node to have access to storage through
PersistentVolumeClaims.

## Motivation

Windows nodes on OpenShift have had persistent storage available through the use of the in-tree storage drivers.
Persistent storage is a necessary feature for production workloads on Kubernetes, and it is important that this
functionality continue to work, even with the in-tree drivers being removed.

### User Stories

* As a developer using OpenShift, I want the Windows workloads I create to be able to read and write to storage which
  persists when pods are removed and re-created.

### Goals

* Windows nodes configured by WMCO fulfill all the pre-requisites to run CSI node drivers.

### Non-Goals

* The Windows CSI storage drivers will not be distributed as part of OpenShift.
  * At this time, we are not able to ship Windows containers images as all Windows container images are packaged

> with a Windows kernel. Red Hat has a policy to not ship 3rd party kernels for support reasons.

Reviewer 3: Nit: Please line up under `At`

> use [host process container images](https://github.com/microsoft/windows-host-process-containers-base-image),

Reviewer 5: `In the future...` might be better suited for the Alternatives section

however that would require CI and CPaaS pipelines to support building Windows containers.
* Enabling the use of volumes which can be shared between Linux and Windows.

## Proposal

The [windows-machine-config-operator](https://github.com/openshift/windows-machine-config-operator/)(WMCO),
will ensure that csi-proxy is running on the Windows nodes it configures. This will be done through the normal
workflow for adding a service to the expected service list.

> The csi-proxy binary will be baked into the WMCO image. WMCO will copy the binary to each node at configuration

Reviewer 5: Could be useful to call out how we will acquire the binary (built from source i assume)

With the nodes in this state, users are free to follow upstream documentation to deploy the CSI node drivers
of their choice on the Windows nodes in the cluster.

No support from Red Hat will be given for issues occurring with the upstream drivers, as they are not being distributed
as part of OpenShift.

> 

Reviewer 3: Nit: extra line

### Workflow Description

1. The cluster administrator will identify what CSI storage drivers the cluster requires for its workloads.
2. Based on the requirements of the workloads, the cluster admin can follow upstream documentation to deploy the CSI
   node drivers on its Windows nodes.

### API Extensions

N/A

### Risks and Mitigations

* If there are any security issues present in the CSI driver images deployed by the user, they will not be fixed

> by upgrading OCP. It is up to the user to stay informed and update their images as necessary.

Reviewer 3: Nit: please line up under `If`

> * A mismatch between the OpenShift CSI drivers and the user deployed Windows CSI drivers could cause undefined

Reviewer 3: ```suggestion
* A mismatch between the OpenShift Linux CSI drivers and the user deployed Windows CSI drivers could cause undefined
```

  behavior. This is described in the [version skew](#version-skew-strategy) section.

### Drawbacks

The main drawback to this approach is it puts the onus on the cluster administrator. As they will be required

> to deploy and maintain the Windows CSI DaemonSet themselves, and will not be given support for any issues that may
> come up, which do not specifically deal with the csi-proxy Windows service. This may cause an increased load

Reviewer 3: ```suggestion
to deploy and maintain the Windows CSI DaemonSet themselves, and will only be given support for issues with the csi-proxy Windows service. This may cause an increased load
```

on both support and developers, as support issues coming in will need to be properly filtered out based on the root
cause of the problem.

## Design Details

### Test Plan

WMCO already has an extensive end to end test suite. To facilitate testing the additions described in this enhancement,
the e2e tests will be augmented to deploy a CSI node driver DaemonSet across Windows nodes, just as a user would. The
e2e tests would then validate that a Windows pod is able to mount a volume backed by the driver. Initially vSphere
storage will be tested, and other platforms can be tested in the future as desired. The purpose of testing is
to ensure CSI proxy is properly interacting with the Windows instance and Windows pods. Testing this on all platforms

> is not completely necessary.

Reviewer 3: Please remove `Testing this on all platforms is not completely necessary`. It is necessary we are just choosing it do it in the future based on customer demand.

### Graduation Criteria

#### Dev Preview -> Tech Preview

N/A

#### Tech Preview -> GA

This feature is targeted for WMCO 9.0.0 in OpenShift 4.14.

#### Removing a deprecated feature

No action required.

> ### Upgrade / Downgrade Strategy

Reviewer 2: There should be a note how the csi-proxy will be upgraded on OCP upgrade. I guess you already have a process / description for a generic host service upgrade, a link would be appropriate.

Reviewer 2: Expectation is that a grpc connection between a CSI driver and the csi proxy can get interrupted during upgrade. We expect that either the CSI driver re-tries the call or passes the error back to Kubernetes and Kubernetes retries.

[When upgrading cluster versions, in-tree driver Volumes will automatically be updated to use the appropriate CSI node
drivers by the cluster](https://docs.openshift.com/container-platform/4.12/storage/container_storage_interface/persistent-storage-csi-migration.html).

If users are making use of the Linux CSI node drivers on the cluster, and not deploying all parts of storage
themselves, users will have to ensure the following:
Before upgrading, users should check the OCP release notes to determine how the Linux CSI drivers have changed, and
if they need to use a new Windows CSI node driver image. If so, the Windows DaemonSet should be updated once the
cluster upgrade is complete. More on this can be found in the [version skew](#version-skew-strategy) section.

WMCO does not support downgrades.

### Version Skew Strategy

If the CSI driver pods included in OCP are used as the driver controller and Linux node drivers, while the Windows
node drivers use the upstream images, there may be a mismatch. This should be mitigated by the cluster administrator
ensuring that the upstream images they use are as close to a version match as possible with the OCP distributed

> drivers.

Reviewer 2: I'd mention that it is expected that a CSI driver runs with different version on different Linux nodes during CSI driver upgrade. It's nothing new, CSI driver vendors know how to handle it. The period when there is a skew could be just longer on Windows.

drivers.

> To eliminate this issue, the cluster administrator can choose to use an OCP cluster with no CSI drivers installed.

Reviewer 3: You are actually describing two scenarios:
- One where the customer needs to use Linux and Windows storage
- And another where the customer needs to use only Windows storage
Please describe it in that manner rather than eliminating  the issue of needing Linux and Windows storage. This gets further confusing when you say `This is only an option if they only plan on using storage on Windows,`

They would then install both the Windows and Linux CSI pods and required resources themselves, using upstream
documentation. This is only an option if they only plan on using storage on Windows, as doing this will prevent the

> user from getting support from Red Hat for any storage issues they may have, both Linux and Windows. With that in mind,

Reviewer 3: ```suggestion
user from getting support from Red Hat for any storage issues they may have on both Linux and Windows. With that in mind,
```

storage will be tested, and other platforms can be tested in the future as desired. The purpose of testing is
to ensure CSI proxy is properly interacting with the Windows instance and Windows pods.

### Graduation Criteria

#### Dev Preview -> Tech Preview

N/A

#### Tech Preview -> GA

This feature is targeted for WMCO 9.0.0 in OpenShift 4.14.

#### Removing a deprecated feature

No action required.

### Upgrade / Downgrade Strategy

#### Cluster upgrades

[When upgrading cluster versions, in-tree driver Volumes will automatically be updated to use the appropriate CSI node
drivers by the cluster](https://docs.openshift.com/container-platform/4.12/storage/container_storage_interface/persistent-storage-csi-migration.html).
In order to properly make use of this, and ensure a clean transition with minimal workload interruption, the CSI node
drivers must be running on the Windows node before the upgrade commences. As the node drivers require csi-proxy to run,
as this feature is targeted for WMCO 9.0.0, a new 8.y.z version of WMCO must be released which runs csi-proxy on each

> Windows node. If this feature is backported to WMCO x.y.z, the same must be done for WMCO x-1.y.z. This will allow the

Reviewer 3: >If this feature is backported to WMCO x.y.z, the same must be done for WMCO x-1.y.z.

Is that true? This implies we have to backport CSI-proxy to all supported WMCO versions.

Author: I'm not sure I follow, if we decided to backport the out-of-tree storage feature to 4.12, this only means we have to backport CSI proxy to 4.11.
